File: mobile/frontend/lib/data/repositories/product_repository.dart
import 'package:frontend/data/models/product.dart';
import 'base_repository.dart';

class ProductRepository extends BaseRepository<Product> {
  ProductRepository() : super('products');

  @override
  Product fromMap(Map<String, dynamic> map) => Product.fromMap(map);

  @override
  Map<String, dynamic> toMap(Product entity) => entity.toMap();

  // ── Domain queries ────────────────────────────────────────

  /// Find product by SKU code.
  Future<Product?> findBySku(String sku) => findOne('sku', sku);

  /// Get active products only.
  Future<List<Product>> getActiveProducts() {
    return query(where: 'actif = 1', orderBy: 'nom_produit ASC');
  }

  /// Get all products sorted by name.
  Future<List<Product>> getAllSorted() {
    return getAll(orderBy: 'nom_produit ASC');
  }

  /// Get products by category.
  Future<List<Product>> getByCategory(String category) {
    return query(where: 'categorie = ?', whereArgs: [category]);
  }

  /// Search products by name or SKU.
  Future<List<Product>> search(String term) {
    final like = '%$term%';
    return query(
      where: '(nom_produit LIKE ? OR sku LIKE ?)',
      whereArgs: [like, like],
      orderBy: 'nom_produit ASC',
    );
  }

  /// Get distinct categories.
  Future<List<String>> getCategories() async {
    final db = await database;
    final results = await db.rawQuery(
      'SELECT DISTINCT categorie FROM products WHERE categorie IS NOT NULL AND is_deleted = 0 ORDER BY categorie ASC',
    );
    return results.map((r) => r['categorie'] as String).toList();
  }

  /// Toggle product active status.
  Future<void> toggleActive(String productId) async {
    final product = await getById(productId);
    if (product != null) {
      await update(productId, {'actif': product.isActive ? 0 : 1});
    }
  }
}

File: mobile/backend/app/repositories/product_repository.py
"""
Product repository for Firestore operations.
"""

from typing import List, Optional, Dict, Any

from app.repositories.base_repository import BaseRepository


class ProductRepository(BaseRepository):
    """Repository for Product documents."""

    def __init__(self):
        super().__init__("products")

    async def get_by_sku(self, sku: str) -> Optional[Dict[str, Any]]:
        """Find a product by SKU."""
        return await self.find_one("sku", sku)

    async def get_by_category(self, category: str) -> List[Dict[str, Any]]:
        """Get all products in a category."""
        return await self.query(filters=[("categorie", "==", category)])

    async def get_active_products(self) -> List[Dict[str, Any]]:
        """Get all active products."""
        return await self.query(filters=[("actif", "==", True)])

    async def get_filtered(
        self, category: Optional[str] = None, active_only: bool = False
    ) -> List[Dict[str, Any]]:
        """Get products with optional filters."""
        filters = []
        if category:
            filters.append(("categorie", "==", category))
        if active_only:
            filters.append(("actif", "==", True))
        return await self.query(filters=filters if filters else None)

    async def get_high_demand(self, threshold: float = 1.0) -> List[Dict[str, Any]]:
        """Get products with demand_freq above threshold."""
        return await self.query(filters=[("demand_freq", ">=", threshold)])
